cpp/src/copying/scatter.cu

```diff
@@ -0,0 +1,383 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/copying.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/scatter.hpp>
+#include <cudf/detail/gather.cuh>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/detail/utilities/cuda.cuh>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+namespace {
+
+void set_null_mask(std::unique_ptr<column>& col, rmm::device_buffer&& bitmask,
+    size_type null_count)
+{
+  auto const size = col->size();
+  auto const type = col->type();
+  auto contents = col->release();
+  col = std::make_unique<column>(type, size, std::move(*contents.data),
+    std::move(bitmask), null_count, std::move(contents.children));
+  // TODO use this from PR 3172 instead
+  //copy->set_null_mask(std::move(mask), null_count);
+}
+
+template <typename T, typename MapIterator>
+thrust::device_vector<T> make_gather_map(MapIterator scatter_iter,
+    size_type scatter_rows, size_type gather_rows, cudaStream_t stream)
+{
+  static_assert(std::is_signed<T>::value,
+    "Need different invalid index if unsigned index types are added");
+  auto const invalid_index = static_cast<T>(-1);
+
+  // Convert scatter map to a gather map
+  auto gather_map = thrust::device_vector<T>(gather_rows, invalid_index);
```

Comment:
Why not rmm::device_vector ?

Reply:
Ok, replaced rmm::device_vector everywhere